tools/kani-cov/src/coverage.rs

```diff
@@ -0,0 +1,209 @@
+    }
+}
+
+/// Raw Kani coverage results.
+///
+/// Note: This data structure should not be duplicated in Kani -
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CoverageResults {
+    pub data: BTreeMap<String, Vec<CoverageCheck>>,
+}
+
+/// Aggregated coverage results.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CombinedCoverageResults {
+    pub data: BTreeMap<String, Vec<(String, Vec<CovResult>)>>,
+}
+
+/// The coverage result associated to a particular coverage region.
+/// Basically, this aggregates the information of one or more `CoverageCheck`.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CovResult {
+    pub function: String,
+    pub region: CoverageRegion,
+    pub times_covered: u32,
+    pub total_times: u32,
+}
+
+/// A coverage region.
+///
+/// Note: This data structure should not be duplicated in Kani -
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct CoverageRegion {
+    pub file: String,
+    pub start: (u32, u32),
+    pub end: (u32, u32),
+}
+
+impl Display for CoverageRegion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{} - {}:{}", self.start.0, self.start.1, self.end.0, self.end.1)
+    }
+}
+
+/// A coverage term.
+///
+/// Note: This data structure should not be duplicated in Kani -
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum CoverageTerm {
+    Counter(u32),
+    Expression(u32),
+}
+
+/// The coverage information to produce for a particular file.
+pub struct FileCoverageInfo {
+    pub filename: String,
+    pub function: CoverageMetric,
+    pub line: CoverageMetric,
+    pub region: CoverageMetric,
+}
+
+/// A coverage metric.
+pub struct CoverageMetric {
+    pub covered: u32,
+    pub total: u32,
+}
+
+impl CoverageMetric {
+    pub fn new(covered: u32, total: u32) -> Self {
+        CoverageMetric { covered, total }
+    }
+}
+
+/// Function information obtained through a tree-sitter
+#[derive(Debug)]
+pub struct FunctionInfo {
+    pub name: String,
+    pub start: (usize, usize),
+    pub end: (usize, usize),
+}
+
+/// Extract function information from a file using a tree-sitter
+pub fn function_info_from_file(filepath: &PathBuf) -> Vec<FunctionInfo> {
+    let source_code = fs::read_to_string(filepath).expect("could not read source file");
+    let mut parser = Parser::new();
+    parser.set_language(&tree_sitter_rust::language()).expect("Error loading Rust grammar");
+
+    let tree = parser.parse(&source_code, None).expect("Failed to parse file");
+
+    let mut cursor = tree.walk();
+    let first_child_exists = cursor.goto_first_child();
+
+    if !first_child_exists {
+        return vec![];
+    }
+
+    let mut function_info: Vec<FunctionInfo> = Vec::new();
+
```

Comment:
We could also declare here:
```rust
let source_bytes = source_code.as_bytes();
```
and avoid calling it multiple times.

Reply:
Yep, done.